Solard idea

pigmej opened the pull request:
Design:
Currently it uses pure TCP sockets. I have implemented TCP + msgpack based protocol which can work as rpc (run transport) and bidireactional stream (files). It could support any protocol, I added some unified interface on client and server sides so later we can use "anything" we want. For now there is one client transport for `True`/`False` `use_sudo`. (transparent)

Currently TCP server forks when it requires other user permission, so it should be started from privileged user (or no user switching possible).

Functions:
- it is capable to work as run transport
- it is capable to work as sync transport

More functions:
- we can integrate torrent client with it
- we can integrate "remote python file execution" with it (current ansible_playbook for example).


For further works we need solve "transports as resource" problem first.

dshulyak commented on solard/setup.py:
```diff
@@ -0,0 +1,51 @@
+    keywords='deployment',
+    packages=find_packages(),
+    zip_safe=False,
+    # install_requires=find_requires(),
```
msgpack, gevent ?

pigmej replied:
Added :)

dshulyak commented on solard/solard/tcp_server.py:
```diff
@@ -0,0 +1,291 @@
+    #     self.sock.sendall(data)
+
+    def _write_gen_end(self):
+        data = {'st': 21, 'res': None}
```
this codes - 'st' - it will be like our convention? if so consts or at least message would be nice

pigmej replied:
It's the type of response, I can make them consts obviously :)

pigmej replied:
Addressed that in 345f487

pigmej commented:
So I changed:
* removed gevent (threads + fork)
* solard runs with start-stop-daemon now
* added BAT transport (Best Available Transport) which assumes that transports are in that order: `solard`, `rsync`, `ssh`.
* Changed / improved location_id/transports_id
* Added solard example (`examples/solard/example.py`)

pigmej commented:
Easiest way of testing:
`python examples/riak/riaks.py deploy`
`python examples/riak/riaks.py add_solard {node_num}`
.. normal solar run stuff

It will make solard transport available on `{node_num}`, and all actions for riak will be excecuted through solard.

dshulyak commented:
works for me, tried on clean vagrant boxes

+1
